rfc/graph-feature-flags.md

```diff
@@ -0,0 +1,242 @@
+
+GFF allows you to publish Subgraph Schema Changes under a feature flag.
+The runtime can dynamically enable or disable feature flags for each request.
+
+## Motivation
+
+There are four main reasons, and the correlated problems to solve, why we want to add Graph Feature Flags to Open Federation:
+
+1. **Feature Rollout**: Feature flags allow you to roll out new features gradually.
+   You can enable a feature for a small percentage of users and monitor its performance.
+   If everything goes well, you can increase the percentage of users who see the feature.
+   If something goes wrong, you can disable the feature without deploying a new version of the API.
+2. **Schema Evolution**: Feature flags allow you to publish Subgraph Schema Changes without breaking existing clients.
+   You can add new fields or types to your schema under a feature flag.
+   Existing clients can continue to use the old schema until they are ready to switch to the new schema.
+   New clients will automatically use the latest version of the schema, including all feature flags.
+3. **Dynamic Configuration**: Feature flags allow you to change the behavior of your API at runtime.
+   You can enable or disable feature flags for each request based on the user's permissions or other criteria.
+   This allows you to create personalized experiences for your users without deploying a new version of the API.
+4. **Staging Environments**: Feature flags allow you to test new features in a staging environment before rolling them out to production.
+   You can enable a feature flag in the staging environment to test it with real data and real users.
+   If everything goes well, you can enable the feature flag in the production environment.
+   If something goes wrong, you can disable the feature flag in the staging environment without affecting the production environment.
+
+Let's break down each of these points in more detail.
+
+### Feature Rollout
+
+Feature flags allow you to roll out new features gradually.
+Let's say you're starting with a monolithic GraphQL API and want to migrate to a Federated GraphQL API.
+You've got existing clients that depend on the old schema and have certain expectations about the characteristics of the API,
+like the performance of certain GraphQL Operations.
+
+With feature flags, we can publish a new Subgraph Schema that `@overrides` a field in the old schema, but don't automatically enable it for all clients.
+Instead, we can send a small percentage of traffic to the new Subgraph in a "shadow mode" and monitor its performance.
+This means that the new Subgraph is used to compute the result of the GraphQL Operation, but the result is not returned to the client.
+Instead, we compare the result of the newly introduced Subgraph with the result of the existing Monolith in terms of correctness and performance.
+
+If everything goes well, we can enable the feature flag for 1% of your users, keep monitoring,
```

Comment:
How do you want to implement this? A router fleet won't be able to serve an exact percentage of the traffic because of a split-brain scenario. In the simplest implementation, it would be a rough estimate. Based on the load-balancing algorithm of the LB (least-connections), it could diverge even further. Usually, this is a task of a ingress or LB.

Both approaches has benefits:

**Proxy**
- Almost every proxy like Nginx, Caddy, Traefik can gradually roll-out traffic based on the feature flag header, Session affinity, Cookie. This integrates with existing platform / monitoring tools.
- Requires to deploy a different router instance for every feature flag which is quite a burden but can be automated.

**Router**
- We can test / rollout feature flags without deploying additional routers.
- Cosmo would need to provide excellent integration with OTEL / Prometheus to make gradual roll-outs approachable, from an operational and observability standpoint. Cosmo Studio needs to be super well integrated with feature-flag across the board.

In summary, I think the complexity we would hold back from the customer with an integrated solution will be huge. When we are doing it right, it can enable small companies very fast to start experiments. Bigger companies could use Cosmo or build upon this foundation to provide GraphQL specific tools. The tooling and integration are the keys here, but this is feasible.